[Tests] Occasional failures because of "still open connections" in MockTransportService.doClose

On teardown of the MockTransportService used in unit tests we check that all connections that were opened are closed correctly again. This seems to fail occasionally with errors that look like:

```
  1> [2017-01-25T22:30:19,038][INFO ][o.e.a.s.RemoteClusterConnectionTests] [testNodeDisconnected]: after test
  2> REPRODUCE WITH: gradle :core:test -Dtests.seed=8A5ADED8D62B415D -Dtests.class=org.elasticsearch.action.search.RemoteClusterConnectionTests -Dtests.method="testNodeDisconnected" -Dtests.security.manager=true -Dtests.locale=ru -Dtests.timezone=America/Tijuana
FAILURE 0.12s J0 | RemoteClusterConnectionTests.testNodeDisconnected <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: still open connections: {{BtOlW7kRTK6YUTDmeeI-cg}{uiX1VgzsQmmqjLTVuqH2HQ}{127.0.0.1}{127.0.0.1:9301}=[org.elasticsearch.test.transport.MockTransportService$5@2cc22be3]}
   > 	at __randomizedtesting.SeedInfo.seed([8A5ADED8D62B415D:430ECAC96D5DA0DF]:0)
   > 	at org.elasticsearch.test.transport.MockTransportService.doClose(MockTransportService.java:763)
   > 	at org.elasticsearch.common.component.AbstractLifecycleComponent.close(AbstractLifecycleComponent.java:106)
   > 	at org.elasticsearch.action.search.RemoteClusterConnectionTests.testNodeDisconnected(RemoteClusterConnectionTests.java:212)
   > 	at java.lang.Thread.run(Thread.java:745)
```
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+periodic/1516/console

Unfortunately I cannot reproduce with the given seed, so just opening this to track this or similar build failures for now.

--- a/core/src/main/java/org/elasticsearch/action/search/RemoteClusterConnection.java
+++ b/core/src/main/java/org/elasticsearch/action/search/RemoteClusterConnection.java
@@ -370,6 +370,9 @@ final class RemoteClusterConnection extends AbstractComponent implements Transpo
                             ClusterStateRequest request = new ClusterStateRequest();
                             request.clear();
                             request.nodes(true);
+                            // here we pass on the connection since we can only close it once the sendRequest returns otherwise
+                            // due to the async nature (it will return before it's actually send) this can cause the request to fail
+                            // due to an already closed connection.
                             transportService.sendRequest(connection,
                                 ClusterStateAction.NAME, request, TransportRequestOptions.EMPTY,
                                 new SniffClusterStateResponseHandler(transportService, connection, listener, seedNodes,
@@ -443,24 +446,30 @@ final class RemoteClusterConnection extends AbstractComponent implements Transpo
             @Override
             public void handleResponse(ClusterStateResponse response) {
                 try {
-                    cancellableThreads.executeIO(() -> {
-                        DiscoveryNodes nodes = response.getState().nodes();
-                        Iterable<DiscoveryNode> nodesIter = nodes.getNodes()::valuesIt;
-                        for (DiscoveryNode node : nodesIter) {
-                            if (nodePredicate.test(node) && connectedNodes.size() < maxNumRemoteConnections) {
-                                try {
-                                    transportService.connectToNode(node, remoteProfile); // noop if node is connected
-                                    connectedNodes.add(node);
-                                } catch (ConnectTransportException | IllegalStateException ex) {
-                                    // ISE if we fail the handshake with an version incompatible node
-                                    // fair enough we can't connect just move on
-                                    logger.debug((Supplier<?>)
-                                        () -> new ParameterizedMessage("failed to connect to node {}", node), ex);
+                    try (Closeable theConnection = connection) { // the connection is unused - see comment in #collectRemoteNodes
+                        // we have to close this connection before we notify listeners - this is mainly needed for test correctness
+                        // since if we do it afterwards we might fail assertions that check if all high level connections are closed.
+                        // from a code correctness perspective we could also close it afterwards. This try/with block will
+                        // maintain the actual exceptions thrown from within the try block and suppress the ones that are possible thrown
+                        // by closing hte connection
+                        cancellableThreads.executeIO(() -> {
+                            DiscoveryNodes nodes = response.getState().nodes();
+                            Iterable<DiscoveryNode> nodesIter = nodes.getNodes()::valuesIt;
+                            for (DiscoveryNode node : nodesIter) {
+                                if (nodePredicate.test(node) && connectedNodes.size() < maxNumRemoteConnections) {
+                                    try {
+                                        transportService.connectToNode(node, remoteProfile); // noop if node is connected
+                                        connectedNodes.add(node);
+                                    } catch (ConnectTransportException | IllegalStateException ex) {
+                                        // ISE if we fail the handshake with an version incompatible node
+                                        // fair enough we can't connect just move on
+                                        logger.debug((Supplier<?>)
+                                            () -> new ParameterizedMessage("failed to connect to node {}", node), ex);
+                                    }
                                 }
                             }
-                        }
-                    });
-                    connection.close();
+                        });
+                    }
                     listener.onResponse(null);
                 } catch (CancellableThreads.ExecutionCancelledException ex) {
                     listener.onFailure(ex); // we got canceled - fail the listener and step out
@@ -469,9 +478,6 @@ final class RemoteClusterConnection extends AbstractComponent implements Transpo
                         () -> new ParameterizedMessage("fetching nodes from external cluster {} failed",
                             clusterAlias), ex);
                     collectRemoteNodes(seedNodes, transportService, listener);
-                } finally {
-                    // just to make sure we don't leak anything we close the connection here again even if we managed to do so before
-                    IOUtils.closeWhileHandlingException(connection);
                 }
             }
 

--- a/core/src/main/java/org/elasticsearch/action/search/RemoteClusterConnection.java
+++ b/core/src/main/java/org/elasticsearch/action/search/RemoteClusterConnection.java
@@ -371,7 +371,7 @@ final class RemoteClusterConnection extends AbstractComponent implements Transpo
                             request.clear();
                             request.nodes(true);
                             // here we pass on the connection since we can only close it once the sendRequest returns otherwise
-                            // due to the async nature (it will return before it's actually send) this can cause the request to fail
+                            // due to the async nature (it will return before it's actually sent) this can cause the request to fail
                             // due to an already closed connection.
                             transportService.sendRequest(connection,
                                 ClusterStateAction.NAME, request, TransportRequestOptions.EMPTY,
@@ -450,8 +450,8 @@ final class RemoteClusterConnection extends AbstractComponent implements Transpo
                         // we have to close this connection before we notify listeners - this is mainly needed for test correctness
                         // since if we do it afterwards we might fail assertions that check if all high level connections are closed.
                         // from a code correctness perspective we could also close it afterwards. This try/with block will
-                        // maintain the actual exceptions thrown from within the try block and suppress the ones that are possible thrown
-                        // by closing hte connection
+                        // maintain the possibly exceptions thrown from within the try block and suppress the ones that are possible thrown
+                        // by closing the connection
                         cancellableThreads.executeIO(() -> {
                             DiscoveryNodes nodes = response.getState().nodes();
                             Iterable<DiscoveryNode> nodesIter = nodes.getNodes()::valuesIt;

